src/native/containers/dn-umap.h

```diff
@@ -0,0 +1,312 @@
+	dn_umap_key_dispose_func_t key_dispose_func,
+	dn_umap_value_dispose_func_t value_dispose_func);
+
+bool
+_dn_umap_init (
+	dn_umap_t *map,
+	dn_allocator_t *allocator,
+	dn_umap_hash_func_t hash_func,
+	dn_umap_equal_func_t equal_func,
+	dn_umap_key_dispose_func_t key_dispose_func,
+	dn_umap_value_dispose_func_t value_dispose_func);
+
+dn_umap_it_t
+_dn_umap_begin (dn_umap_t *map);
+
+bool
+_dn_umap_it_next (dn_umap_it_t *it);
+
+static inline void
+dn_umap_it_advance (
+	dn_umap_it_t *it,
+	uint32_t n)
+{
+	while (n && _dn_umap_it_next (it))
+		n--;
+}
+
+static inline dn_umap_it_t
+dn_umap_it_next (dn_umap_it_t it)
+{
+	_dn_umap_it_next (&it);
+	return it;
+}
+
+static inline void *
+dn_umap_it_key (dn_umap_it_t it)
+{
+	DN_ASSERT (it._internal._node);
+	return it._internal._node->key;
+}
+
+#define dn_umap_it_key_t(it, type) \
+	(type)(uintptr_t)(dn_umap_it_key (it))
+
+static inline void *
+dn_umap_it_value (dn_umap_it_t it)
+{
+	DN_ASSERT (it._internal._node);
+	return it._internal._node->value;
+}
+
+#define dn_umap_it_value_t(it, type) \
+	(type)(uintptr_t)(dn_umap_it_value ((it)))
+
+static inline bool
+dn_umap_it_begin (dn_umap_it_t it)
+{
+	dn_umap_it_t begin = _dn_umap_begin (it._internal._map);
+	return (begin._internal._node == it._internal._node && begin._internal._index == it._internal._index);
+}
+
+static inline bool
+dn_umap_it_end (dn_umap_it_t it)
+{
+	return !(it._internal._node);
+}
+
+#define DN_UMAP_FOREACH_BEGIN(map,key_type,key_name,value_type,value_name) do { \
+	key_type key_name; \
+	value_type value_name; \
+	for (dn_umap_it_t __it##key_name = dn_umap_begin (map); !dn_umap_it_end (__it##key_name); __it##key_name = dn_umap_it_next (__it##key_name)) { \
+		key_name = ((key_type)(uintptr_t)dn_umap_it_key (__it##key_name)); \
+		value_name = ((value_type)(uintptr_t)dn_umap_it_value (__it##key_name));
+
+#define DN_UMAP_FOREACH_KEY_BEGIN(map,key_type,key_name) do { \
+	key_type key_name; \
+	for (dn_umap_it_t __it##key_name = dn_umap_begin (map); !dn_umap_it_end (__it##key_name); __it##key_name = dn_umap_it_next (__it##key_name)) { \
+		key_name = ((key_type)(uintptr_t)dn_umap_it_key (__it##key_name));
+
+#define DN_UMAP_FOREACH_END \
+		} \
+	} while (0)
+
+static inline dn_umap_t *
+dn_umap_custom_alloc (
+	dn_allocator_t *allocator,
+	dn_umap_hash_func_t hash_func,
+	dn_umap_equal_func_t equal_func,
+	dn_umap_key_dispose_func_t key_dispose_func,
+	dn_umap_value_dispose_func_t value_dispose_func)
+{
+	return _dn_umap_alloc (allocator, hash_func, equal_func, key_dispose_func, value_dispose_func);
+}
+
+static inline dn_umap_t *
+dn_umap_alloc (void)
+{
+	return _dn_umap_alloc (DN_DEFAULT_ALLOCATOR, NULL, NULL, NULL, NULL);
+}
+
+void
+dn_umap_free (dn_umap_t *map);
+
+static inline bool
+dn_umap_custom_init (
+	dn_umap_t *map,
+	dn_allocator_t *allocator,
+	dn_umap_hash_func_t hash_func,
+	dn_umap_equal_func_t equal_func,
+	dn_umap_key_dispose_func_t key_dispose_func,
+	dn_umap_value_dispose_func_t value_dispose_func)
```

Comment:
I know this is how `g_hash_table_new_full` works, but I wonder if it's worth making a utility struct for all the custom arguments:

```c
  dn_unmap_customize_t c = {0,};
  c.allocator = whatever_allocator;
  c.hash_func = my_hash;
  c.equal_func = my_eq;
  // leaving dispose funcs as null

  dn_unmap_custom_init(my_map, &c); // the callee copies out the fields of dn_unmap_customize_t into its internals; it doesn't store the passed-in struct
```

Reply:
I can convert the the custom alloc/init to accept structs to make the function signatures simpler.